draft-ietf-quic-applicability.md

```diff
@@ -566,42 +607,41 @@ to the server instance. The server can provide an IPv4 and an IPv6 address in a
-{{Section 8.1 of QUIC-HTTP}}).
-
-The CONNECTION_CLOSE frame provides an optional reason field, that can be used
-to append human-readable information to an error code.  RESET_STREAM and
-STOP_SENDING frames also include an error code, but no reason string.
-
-Alternatively, a QUIC connection can be silently closed by each endpoint
-separately after an idle timeout. If enabled as indicated by a transport
-parameter in the handshake, the idle timeout is announced for each endpoint
-during connection establishment and the effective value for this connection is
-the minimum of the two values advertised by client and server. An application
-therefore should be able to configure its own maximum value as well as have
-access to the computed minimum value for this connection. An application may
-adjust the maximum idle timeout for new connections based on the number of open
-or expected connections, since shorter timeout values may free-up memory more
-quickly.
-
-If an application desires to keep the connection open for longer than the
-announced timeout, it can send keep-alive messages; a QUIC implementation may
-provide an option to defer the time-out by sending keep-alive messages at the
-transport layer to avoid unnecessary load, as specified in {{Section 10.1.2 of
-QUIC}}. See {{resumption-v-keepalive}} for further guidance on keep-alives.
+# Connection Termination
+
+QUIC connections are terminated in one of three ways: implicit idle timeout,
+explicit immediate close, or explicit stateless reset.
+
+QUIC does not provide any mechanism for graceful connection termination;
+applications using QUIC can define their own graceful termination process (see,
+for example, {{Section 5.2 of QUIC-HTTP}}).
+
+QUIC idle timeout is enabled via transport parameters. Client and server
+announce a timeout period and the effective value for the connection is the
+minimum of the two values. After the timeout period elapses, the connection is
+silently closed. An application therefore should be able to configure its own
+maximum value, as well as have access to the computed minimum value for this
+connection. An application may adjust the maximum idle timeout for new
+connections based on the number of open or expected connections, since shorter
+timeout values may free-up resources more quickly.
+
+As long as application data is exchanged, the idle timeout does not expire. If
+an application has no data to send but desires to keep the connection open for
+longer than the timeout period, it can send keep-alive messages. A QUIC
+implementation may provide an option to defer the time-out by sending keep-alive
+messages at the transport layer to avoid unnecessary load, as specified in
+{{Section 10.1.2 of QUIC}}. Alternatively, applications using QUIC could define
+their own mechanism, such as an application-layer ping, that achieves a similar
+result. See {{resumption-v-keepalive}} for further guidance on keep-alives.
```

Comment:
I'd say you "only" should use application layer ping if you also need that information for the application layer logic itself. If you do application layer pings that just looks like payload traffic on the transport which means even if QUIC is configured to send keep-alives it will not send any because it never becomes idle. So not sure if we actually need to say anything about application layer pings here at all...?

Reply:
That's already what this paragraph says IMO - if there is no useful application data to send, you can use QUIC keepalives or design an application mechanism. The important thing is that applications might have their own idle timeouts above QUIC (for example, time between HTTP requests, time between DATA frames) and its totally fine for applications to have those and design mechanisms that avoid them being triggered.

Reply:
Yes, the tiny different I would be looking for is to say it more like: if there is no useful application data to send, you should use QUIC keepalives or only design an application mechanism if that is needed on the app layer anyway for something else. Or to put if differently: please don't design your own app layer ping just because you can.

Reply:
I disagree. An application protocol designer probably wants wide interoperability and they could be sorely disappointed to find that QUIC implementations don't all  expose a keep alive mechanism. The best way to mitigate that risk is to pretend transport keepalives don't exist and to always have an application one. Doing it that way, means that all implementations of an application protocol can expect the same behaviour.

The text here attempts to tread a fine line by stating what is possible but without recommending any particular course of action.

Reply:
You are right on the point that this discussion actually doesn't belong in this document as that question is not QUIC specific. Therefore I would still advocate for removing the second to last sentence here about app layer pings because of course applications can always do that but is also not specific for applications that use QUIC.

Reply:
Hmm. So given that this doc is focussed on pointing out caveats for applications, how about we swizzle the text a bit to say 

"Application data exchanged on streams or in datagrams defers the QUIC idle timeout. Applications that provide their own keep-alive mechanisms will therefore keep a QUIC connection alive. Applications that don't provide their own keep-alive might be able to use transport-layer mechanism (see {{Section 10.1.2 of QUIC}}, and {{resumption-v-keepalive}}). However, QUIC implementation interfaces for controlling such transport behaviour can vary, affecting the robustness of such designs.

Reply:
WFM. Thanks!